src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -3656,44 +3656,134 @@ static QVariant fcnMakeRectangleFrom3Points( const QVariantList &values, const Q
   return QVariant::fromValue( QgsGeometry( rect.toPolygon() ) );
 }
 
-static QVariant pointAt( const QVariantList &values, const QgsExpressionContext *context, QgsExpression *parent ) // helper function
+static QVariant pointAt( const QgsGeometry &geom, int idx, QgsExpression *parent ) // helper function
 {
-  FEAT_FROM_CONTEXT( context, f )
-  int idx = QgsExpressionUtils::getNativeIntValue( values.at( 0 ), parent );
-  QgsGeometry g = f.geometry();
-  if ( g.isNull() )
+  if ( geom.isNull() )
     return QVariant();
 
   if ( idx < 0 )
   {
-    idx += g.constGet()->nCoordinates();
+    idx += geom.constGet()->nCoordinates();
   }
-  if ( idx < 0 || idx >= g.constGet()->nCoordinates() )
+  if ( idx < 0 || idx >= geom.constGet()->nCoordinates() )
   {
     parent->setEvalErrorString( QObject::tr( "Index is out of range" ) );
     return QVariant();
   }
+  return QVariant::fromValue( geom.vertexAt( idx ) );
+}
+
+// function used for the old $ style
+static QVariant fcnOldXat( const QVariantList &values, const QgsExpressionContext *context, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  FEAT_FROM_CONTEXT( context, feature )
+  const QgsGeometry geom = feature.geometry();
+  int idx = QgsExpressionUtils::getNativeIntValue( values.at( 0 ), parent );
+
+  QVariant v = pointAt( geom, idx, parent );
+
+  if ( !v.isNull() )
+    return QVariant( v.value<QgsPoint>().x() );
+  else
+    return QVariant();
+}
+static QVariant fcnXat( const QVariantList &values, const QgsExpressionContext *f, QgsExpression *parent, const QgsExpressionNodeFunction *node )
+{
+  if ( values.at( 1 ).isNull() || values.at( 0 ).isNull() ) // the case where the alias x_at function is called like a $ function (x_at(i))
+  {
+    return fcnOldXat( values, f, parent, node );
+  }
+  else if ( values.at( 0 ).isNull() && !values.at( 1 ).isNull() ) // same as above with x_at(i:=0) (this values is at the second position)
+  {
+    return fcnOldXat( QVariantList() << values[1], f, parent, node );
+  }
+
+  QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  int vertexNumber = QgsExpressionUtils::getNativeIntValue( values.at( 1 ), parent ); // Should be one or 0 ???
```

Comment:
I forgot to remove it... I write that because some functions start index at 0 and someones at 1. See above https://github.com/qgis/QGIS/pull/50853#issuecomment-1310483776.